tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs

```diff
@@ -0,0 +1,351 @@
+
+type Thing() =
+    member val Do: System.Nullable<float> = System.Nullable() with get,set
+    
+let test() = Thing(Do = 100)
+        """,
+            (fun verifier -> verifier.VerifyIL [
+            """
+  .method public static class Test/Thing 
+          test() cil managed
+  {
+    
+    .maxstack  4
+    .locals init (class Test/Thing V_0)
+    IL_0000:  newobj     instance void Test/Thing::.ctor()
+    IL_0005:  stloc.0
+    IL_0006:  ldloc.0
+    IL_0007:  ldc.i4.s   100
+    IL_0009:  conv.r8
+    IL_000a:  newobj     instance void valuetype [runtime]System.Nullable`1<float64>::.ctor(!0)
+    IL_000f:  callvirt   instance void Test/Thing::set_Do(valuetype [runtime]System.Nullable`1<float64>)
+    IL_0014:  ldloc.0
+    IL_0015:  ret
+  }
+            """
+            ]))
+
+    [<Test>]
+    let ``int converts to System.Nullable<int> in method call property setter``() =
+        CompilerAssert.CompileLibraryAndVerifyILWithOptions([|"--optimize-"|],
+            """
+module Test
+
+type Thing() =
+    member val Do: System.Nullable<int> = System.Nullable() with get,set
+    
+let test() = Thing(Do = 100)
+        """,
+            (fun verifier -> verifier.VerifyIL [
+            """
+  .method public static class Test/Thing 
+          test() cil managed
+  {
+    
+    .maxstack  4
+    .locals init (class Test/Thing V_0)
+    IL_0000:  newobj     instance void Test/Thing::.ctor()
+    IL_0005:  stloc.0
+    IL_0006:  ldloc.0
+    IL_0007:  ldc.i4.s   100
+    IL_0009:  newobj     instance void valuetype [runtime]System.Nullable`1<int32>::.ctor(!0)
+    IL_000e:  callvirt   instance void Test/Thing::set_Do(valuetype [runtime]System.Nullable`1<int32>)
+    IL_0013:  ldloc.0
+    IL_0014:  ret
+  } 
+            """
+            ]))
+
+    [<Test>]
+    let ``int converts to System.Nullable<int> in method call parameter``() =
+        CompilerAssert.CompileLibraryAndVerifyILWithOptions([|"--optimize-"|],
+            """
+module Test
+
+type Thing() =
+    static member Do(i: System.Nullable<int>) = ()
+    
+let test() = Thing.Do(100)
+        """,
+            (fun verifier -> verifier.VerifyIL [
+            """
+  .method public static void  test() cil managed
+  {
+    
+    .maxstack  8
+    IL_0000:  ldc.i4.s   100
+    IL_0002:  newobj     instance void valuetype [runtime]System.Nullable`1<int32>::.ctor(!0)
+    IL_0007:  call       void Test/Thing::Do(valuetype [runtime]System.Nullable`1<int32>)
+    IL_000c:  ret
+  } 
+            """
+            ]))
+
+    [<Test>]
+    let ``Passing an incompatible argument for System.Nullable<'T> method call parameter produces accurate error``() =
+        CompilerAssert.TypeCheckSingleError
+            """
+module Test
+
+type Thing() =
+    static member Do(i: System.Nullable<float>) = ()
+    
+let test() = Thing.Do(true)
+            """
+            FSharpDiagnosticSeverity.Error
+            193
+            (7, 22, 7, 28)
+            """Type constraint mismatch. The type 
```

Comment:
@dsyme this is a bit of an odd error IMO. 

Last I tested it has to do with the fact we create a type variable (this branch https://github.com/dotnet/fsharp/blob/ffa2e0ddb3f6ed8df0bf91a2e5c2c6a070b8c0f3/src/Compiler/Checking/MethodCalls.fs#L382) that then outputs this on a mismatch instead of `This expression was expected to have type ...`

It's not a big deal but still not the error you'd expect calling a fully concrete method.

Reply:
Yes, I understand.  It's ok - AFAIK we can't remove the creation of that type variable, at least without affecting compat.  This error is OK all in all